android-agent/src/main/java/io/opentelemetry/android/internal/services/applifecycle/AppLifecycleService.kt

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.android.internal.services.applifecycle
+
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.ProcessLifecycleOwner
+import io.opentelemetry.android.internal.services.Startable
+
+/**
+ * This class is internal and not for public use. Its APIs are unstable and can change at any time.
+ */
+class AppLifecycleService internal constructor(
```

Comment:
There's probably a better way to address this, though at the moment the reasoning behind it is that all services are essentially helper tools that might be reused by our own submodules, such as instrumentations. Right now, this service provides app lifecycle info that's useful for the network change instrumentation, and I also think it will provide useful info for the activity/fragment instrumentations too iirc. We could replicate this service's functionality in other modules though, but for now, given that the instrumentations will have a dependency on the core agent anyway, which is where the `AndroidInstrumentation` interface lives, it seemed sensible to reuse those tools.

Reply:
So this is effectively a private interface for "friends" but because of module boundaries, it's not declared internal? I wonder if we can just make it public but exclude it from docs, state it's not a part of the public interface, and is liable to be changed at any time, perhaps even add an annotation in there fore good measure? We're effectively hiding it, just making it hard to use

Reply:
> So this is effectively a private interface for "friends" but because of module boundaries, it's not declared internal?

That's right, this is currently the case for all services. This reminded me of a conversation we had in one of the meetings about potentially making them public for all instrumentations. I think this topic is big enough to have its own conversation, so I've created [this issue](https://github.com/open-telemetry/opentelemetry-android/issues/435) to follow up on it!

Reply:
Yeah, pulling these in an `internal` package, putting comments to say this should not be depended on and can change at any time, and using the annotations like `@hide` to keep it out of auto generated docs is probably the best way of do this given the way Kotlin visibility works. Maybe we can put the implementation and API into different modules so apps will never have to explicitly pull in modules that contain the implementation, but we don't want to overdo this and make it too hard on ourselves either 😅 Definitely something to talk about for later and not now